### PR TITLE
Use OSSALifetimeCompletion in PredictableMemOpt

### PR DESCRIPTION
**Pull request by meg-gupta**
The current algorithm to complete lifetimes is incorrect in a few cases. Use OSSALifetimeCompletion instead.

Fixes rdar://119204768

### CODE REVIEW

**Review by eeckstein — APPROVED** [review]
lgtm

File: lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp [hunk 1]
```diff
@@ -2655,145 +2659,19 @@ bool AllocOptimize::tryToRemoveDeadAllocation() {
...
-              auto loc = RegularLocation::getAutoGeneratedLocation();
-              b.createDestroyValue(loc, v);
-            });
-            return;
-          }
-          }
-          llvm_unreachable("Case that did not return in its body?!");
-        });
+    // enums can have incomplete lifetimes in non payload paths that don't end
+    // in unreachable.
```
Comment (atrick) [inline lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp:2670]: It wasn't immediately obvious what the `forceBoundaryCompletion` flag was for here.
```
Lexical enums can have incomplete lifetimes in non payload paths that don't end
in unreachable. Force their lifetime to end immediately after the last use instead.
```

**Review by atrick — APPROVED** [review]
Nice!

File: lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp (hunk 1 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp]: Added this comment.

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Comment by vanvoorden** [comment]
https://forums.swift.org/t/non-deterministic-5-10-compiler-failures-from-predictablememoryaccessoptimizations-and-swift-5-10/72630

@meg-gupta By any chance was this a fix for what looks like some non-deterministic crashes from 5.10?

**Comment by meg-gupta** [comment]
@vanvoorden This PR fixed an issue that would have been raised by ownership verification. It looks like your crash is somewhere in the type system. Will be helpful if you can provide a reproducer for the issue you are seeing.

**Comment by vanvoorden** [comment]
> It looks like your crash is somewhere in the type system.

@meg-gupta Hmm… looks like https://github.com/apple/swift/pull/71241 and https://github.com/apple/swift/pull/71730 landed in 6.0… would that look related to what I am seeing in 5.10?

> Will be helpful if you can provide a reproducer for the issue you are seeing.

I don't have a clean repro yet… sorry about that. I can keep trying to see if I can find something that crashes consistently in a new package.

**Comment by vanvoorden** [comment]
@meg-gupta I think I might have found a repro. https://github.com/swiftlang/swift/issues/74609. Thanks!

**Pull request merged**